### PR TITLE
Add cross-platform PIC support for static libraries

### PR DESCRIPTION
**Pull request by nirbheek**
With C/C++, on Windows you don't need to pass any arguments for a static library to be PIC. On UNIX platforms you need to pass `-fPIC`.

Other languages such as D have compiler-specific PIC arguments required for PIC support in static libraries on UNIX platforms.

This kwarg allows people to specify which static libraries should be built with PIC support. This is usually used for static libraries that will be linked into shared libraries.

### CODE REVIEW

**Comment by nirbheek** [comment]
The CI failure is in Swift, and I can't reproduce that on my machine. The debug logs don't show the ninja output, so I have no idea what the failure is on the CI. In fact, those tests shouldn't be hitting this codepath at all...

**Comment by jpakkane** [comment]
Yes, we definitely want something like this. While we're at it maybe we should think about -fPIE as well. Should we support it for exes? Static libraries? (though I think you can link -fPIC and -fPIE objects together and it works). The only real use of -fPIE that I know of is Qt 5, which requires it but it puts it in its pkg-config args so it works automatically.

**Comment by jpakkane** [comment]
It just occurred to me that we should check if the user is trying to link a static library that is not set as `pic` into a shared library and print a warning (or error out?) if so.

**Comment by nirbheek** [comment]
I'll look into the PIE stuff. I also had the same idea about static library without pic being linked into a shared library; will implement that.

Also, looks like the Swift failures are happening everywhere, so I'll split that out into its own PR.

**Comment by nirbheek** [comment]
So, I think the `PIE` stuff is a bit more involved. For instance, it will apply to executables and shared libraries too, so I'll do a separate PR for it. This PR now implements everything else.

Of note is that `-fPIC` is also the default on OS X and there is no way to turn it off, so the code reflects this now.

File: test cases/common/62 exe static shared/subdir/exports.h [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#ifndef __MESON_EXPORTS__
```
Comment (jpakkane) [inline test cases/common/62 exe static shared/subdir/exports.h]: The C and C++ standards reserve all preprocessor tokens that start with one or two underscores to the stdlib, so this is undefined behaviour. Please rename or, better yet, use `pragma once`.

File: test cases/failing/33 exe static shared/shlib2.c [hunk 2]
```diff
@@ -0,0 +1,21 @@
+#ifndef __MESON_EXPORTS__
+#define __MESON_EXPORTS__
```
Comment (jpakkane) [inline test cases/failing/33 exe static shared/shlib2.c]: Same here.

**Pull request merged**